---
title: "BSSeq_smoothing"
author: "Shruti Singh Kakan"
date: "2025-03-26"
output: html_document
---

```{r setup, include=FALSE}

#For operating this script on Sherlock, Library installations must be done in Unix Shell. 
knitr::opts_chunk$set(echo = TRUE)

```

#### Loading libraries and other dependencies
```{r Loading dependencies, message=FALSE, warning=FALSE, paged.print=FALSE}
.libPaths('/home/users/singhkak/R/x86_64-pc-linux-gnu-library/4.2')
#library(bsseq,lib.loc='~/R/3.6/library')
library(rtracklayer, lib.loc = '~/R/x86_64-pc-linux-gnu-library/4.2')
library(bsseq)
library(BSgenome.Mmusculus.UCSC.mm10)
#library(BSgenome.Hsapiens.UCSC.hg38)
library(BiocParallel)
library(parallel)

options(MulticoreParam=MulticoreParam(workers=4))
```

#### Chromosome information
```{r}
chr_size <- seqlengths(BSgenome.Mmusculus.UCSC.mm10)
#Removing alternative loci and haplotypes
chr_size <- chr_size[1:22]
#names(chr_size) <- gsub('chr','',names(chr_size))
#names(chr_size)[names(chr_size)=='M'] <- 'MT'
```

### 1. WT Mouse whole Embryo E8.5 dataset (PRJNA541237)
```{r Read in WT Embryo dataset}

names(chr_size) <- gsub('chr','',names(chr_size))
names(chr_size)[names(chr_size)=='M'] <- 'MT'

bsseq_1.1 <- read.bismark(files = c('/scratch/groups/ximenac/PRJNA541237/Bismark_Aligned/Local/Meth_Extract/SRR9016926_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz'), 
                                  colData = DataFrame(row.names = c("SRR9016926")),
                                  rmZeroCov = FALSE, 
                                  strandCollapse = TRUE, 
                                  verbose = TRUE)

bsseq_1.2 <- read.bismark(files = c('/scratch/groups/ximenac/PRJNA541237/Bismark_Aligned/Local/Meth_Extract/SRR11806587_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz'), 
                                  colData = DataFrame(row.names = c("SRR1180658")),
                                  rmZeroCov = FALSE, 
                                  strandCollapse = TRUE, 
                                  verbose = TRUE)

bsseq_1 <- list(bsseq1.1=bsseq_1.1, bsseq1.2=bsseq_1.2)
rm(bsseq_1.1, bsseq_1.2)

for (i in 1:length(bsseq_1)) {
  bsseq_1[[i]] <- sort(bsseq_1[[i]])
  chr_size1 <- chr_size[match(names(seqlengths(bsseq_1[[i]])),names(chr_size))[!is.na(match(names(seqlengths(bsseq_1[[i]])),names(chr_size)))]]
  bsseq_1[i] <- bsseq_1[[i]][seqnames(bsseq_1[[i]]) %in% names(chr_size1),]
  seqlevels(bsseq_1[[i]]) <- names(chr_size1)
  seqlengths(bsseq_1[[i]]) <- chr_size1
}
#Getting Methylation coverage
regions <- GRanges(seqnames = c("18", "19"), 
                   ranges = IRanges(start = 3.9 * 10^7 + c(0,400000), 
                                    width = 1000))
getCoverage(bsseq_1[[i]], regions = regions, what = "perBase")

saveRDS(bsseq_1, ('/home/groups/ximenac/XCD_BSSeq/PRJNA541237/PRJNA541237.bsseq.rds'))

```

### 2. Mouse Embryo Retina Dataset E11.5, E12.5 (PRJNA521997)
```{r}
names(chr_size) <- gsub('chr','',names(chr_size))
names(chr_size)[names(chr_size)=='M'] <- 'MT'

#Create an empty list to save files in
bsseq_2.names <- c(paste0(rep("bsseq_2.", 9), 1:9))
bsseq_2 <- vector("list", length(bsseq_2.names))
names(bsseq_2) <- bsseq_2.names

#Generate file paths
file_paths = list()
for (i in 1:length(bsseq_2)) {
  file_paths[[i]] = c(paste0("/scratch/groups/ximenac/PRJNA521997/Bismark_Aligned/Local/Meth_Extract/SRR856864", i, "_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz"))
}
#Read in files into a bsseq object (This will take about 5-10 minutes)
for (i in 1:length(bsseq_2)) {
  bsseq_2[[i]] <- read.bismark(files = file_paths[[i]],
                               colData = DataFrame(row.names = paste0(("SRR856864"), i)) ,
                                  rmZeroCov = FALSE, 
                                  strandCollapse = TRUE, 
                                  verbose = TRUE)
}

for (i in 1:length(bsseq_2)) {
  bsseq_2[[i]] <- sort(bsseq_2[[i]])
  chr_size1 <- chr_size[match(names(seqlengths(bsseq_2[[i]])),names(chr_size))[!is.na(match(names(seqlengths(bsseq_2[[i]])),names(chr_size)))]]
  bsseq_2[[i]] <- bsseq_2[[i]][seqnames(bsseq_2[[i]]) %in% names(chr_size1),]
  seqlevels(bsseq_2[[i]]) <- names(chr_size1)
  seqlengths(bsseq_2[[i]]) <- chr_size1
}

#Getting Methylation coverage to test if data that was read in is accurate
regions <- GRanges(seqnames = c("18", "19"), 
                   ranges = IRanges(start = 3.9 * 10^7 + c(0,400000), 
                                    width = 1000))
getCoverage(bsseq_2[[3]], regions = regions, what = "perBase")
bsseq_2
saveRDS(bsseq_2, ('/home/groups/ximenac/XCD_BSSeq/PRJNA521997/PRJNA521997.bsseq.rds'))
```

### 3. Embryo & Neonate Retinae E14.5, E17.5, P3 (PRJNA343691)
```{r Reading in and sorting PRJNA34 dataset}
names(chr_size) <- gsub('chr','',names(chr_size))
names(chr_size)[names(chr_size)=='M'] <- 'MT'

bsseq_3.names <- c(paste0(rep("bsseq_3.", 3), 1:3))
bsseq_3 <- vector("list", length(bsseq_3.names))
names(bsseq_3) <- bsseq_3.names
print("Empty list")
bsseq_3
count <- 1
file_paths = list()
for (i in c(699, 700, 702)) {
  file_paths[[count]] = c(paste0("/scratch/groups/ximenac/PRJNA343691/Bismark_Aligned/Local/Meth_Extract/SRR4254", i, "_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz"))
  count <- count+1
}
count <- 1
for (i in c(699, 700, 702)) {
  bsseq_3[[count]] <- read.bismark(files = c(file_paths[[count]]),
                               colData = DataFrame(row.names = paste0(("SRR4254"), i)) ,
                                  rmZeroCov = FALSE, 
                                  strandCollapse = TRUE, 
                                  verbose = TRUE)
  count <- count+1
}

for (i in 1:length(bsseq_3)) {
  bsseq_3[[i]] <- sort(bsseq_3[[i]])
  chr_size1 <- chr_size[match(names(seqlengths(bsseq_3[[i]])),names(chr_size))[!is.na(match(names(seqlengths(bsseq_3[[i]])),names(chr_size)))]]
  bsseq_3[[i]] <- bsseq_3[[i]][seqnames(bsseq_3[[i]]) %in% names(chr_size1),]
  seqlevels(bsseq_3[[i]]) <- names(chr_size1)
  seqlengths(bsseq_3[[i]]) <- chr_size1
}
#Getting Methylation coverage to test if data that was read in is accurate
regions <- GRanges(seqnames = c("18", "19"), 
                   ranges = IRanges(start = 3.9 * 10^7 + c(0,400000), 
                                    width = 1000))
getCoverage(bsseq_3[[3]], regions = regions, what = "perBase")
bsseq_3
saveRDS(bsseq_3, ('/home/groups/ximenac/XCD_BSSeq/PRJNA343691/PRJNA343691.bsseq.rds'))
```

### 4. 10 week old mice Rods (N=2) and Cones (N=2) (PRJNA294311)
```{r Reading in and sorting samples from PRJNA29 dataset}
chr_size <- seqlengths(BSgenome.Mmusculus.UCSC.mm10)
#Removing alternative loci and haplotypes
chr_size <- chr_size[1:22]

bsseq_4.names <- c(paste0(rep("bsseq_4.", 4), 1:4))
bsseq_4 <- vector("list", length(bsseq_4.names))
names(bsseq_4) <- bsseq_4.names
print("Empty list")
bsseq_4
count <- 1
file_paths = list()
for (i in c(46, 47, 50, 51)) {
  file_paths[[count]] = c(paste0("/scratch/groups/ximenac/PRJNA294311/Bismark_Aligned/Local/Meth_Extract/SRR27228", i, "_bismark_bt2.deduplicated.CpG_report.txt.gz"))
  count <- count+1
}
count <- 1
for (i in c(46, 47, 50, 51)) {
  bsseq_4[[count]] <- read.bismark(files = c(file_paths[[count]]),
                               colData = DataFrame(row.names = paste0(("SRR27228"), i)) ,
                                  rmZeroCov = FALSE, 
                                  strandCollapse = TRUE, 
                                  verbose = TRUE)
  count <- count+1
}
for (i in 1:length(bsseq_4)) {
  bsseq_4[[i]] <- sort(bsseq_4[[i]])
  chr_size1 <- chr_size[match(names(seqlengths(bsseq_4[[i]])),names(chr_size))[!is.na(match(names(seqlengths(bsseq_4[[i]])),names(chr_size)))]]
  bsseq_4[[i]] <- bsseq_4[[i]][seqnames(bsseq_4[[i]]) %in% names(chr_size1),]
  seqlevels(bsseq_4[[i]]) <- names(chr_size1)
  seqlengths(bsseq_4[[i]]) <- chr_size1
}

#Getting Methylation coverage to test if data that was read in is accurate
regions <- GRanges(seqnames = c("18", "19"), 
                   ranges = IRanges(start = 3.9 * 10^7 + c(0,400000), 
                                    width = 1000))
getCoverage(bsseq_4[[3]], regions = regions, what = "perBase")
bsseq_4
saveRDS(bsseq_4, ('/home/groups/ximenac/XCD_BSSeq/PRJNA294311/PRJNA294311.bsseq.rds'))
```


### 5. 12 week Rod photoreceptor datasets (GSE134873)
```{r Reading in samples from GSE dataset}
chr_size <- seqlengths(BSgenome.Mmusculus.UCSC.mm10)
#Removing alternative loci and haplotypes
chr_size <- chr_size[1:22]

#Mouse Rod Photo-receptors dataset
bsseq_5.names <- c(paste0(rep("bsseq_5.", 3), 1:3))
bsseq_5 <- vector("list", length(bsseq_5.names))
names(bsseq_5) <- bsseq_5.names
print("Empty list")
bsseq_5
count <- 1
file_paths = list()
for (i in c(2:4)) {
  file_paths[[count]] = c(paste0("/scratch/groups/ximenac/GSE134873/Bismark_Aligned/Local/Meth_Extract/SRR983366", i, "_1_val_1_val_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz"))
  count <- count+1
}
#Reading in data can take up to 2-3 minutes
count <- 1
for (i in c(2:4)) {
  bsseq_5[[count]] <- read.bismark(files = c(file_paths[[count]]),
                               colData = DataFrame(row.names = paste0(("SRR983366"), i)) ,
                                  rmZeroCov = FALSE, 
                                  strandCollapse = TRUE, 
                                  verbose = TRUE)
  count <- count+1
}

#Reducing BSseq object to 19 chromosomes, X, Y and MT
for (i in 1:length(bsseq_5)) {
  bsseq_5[[i]] <- sort(bsseq_5[[i]])
  chr_size1 <- chr_size[match(names(seqlengths(bsseq_5[[i]])),names(chr_size))[!is.na(match(names(seqlengths(bsseq_5[[i]])),names(chr_size)))]]
  bsseq_5[[i]] <- bsseq_5[[i]][seqnames(bsseq_5[[i]]) %in% names(chr_size1),]
  seqlevels(bsseq_5[[i]]) <- names(chr_size1)
  seqlengths(bsseq_5[[i]]) <- chr_size1
}

for (i in 1:length(bsseq_5)) {
  seqlevels(bsseq_5[[i]]) <- seqlevels(bsseq_3[[1]])  # align seqlevels
  seqinfo(bsseq_5[[i]]) <- seqinfo(bsseq_3[[1]])  # guaranteed to work to change "chr1" to "1"
  #print(head(seqnames(bsseq_5[[i]]), n = 4))
  #print(dim(bsseq_5[[i]]))
}

#Getting Methylation coverage
regions <- GRanges(seqnames = c("18", "19"), 
                   ranges = IRanges(start = 3.9 * 10^7 + c(0,400000), 
                                    width = 1000))
getCoverage(bsseq_5[[2]], regions = regions, what = "perBase")

saveRDS(bsseq_5, ('/home/groups/ximenac/XCD_BSSeq/GSE134873/GSE134873.bsseq.rds'))
```


###6. HSC Dataset (PRJNA214817)
```{r}
#Mouse bone Hematopoietic Stem Cells
bsseq_6.names <- c(paste0(rep("bsseq_6", 7), 1:7))
bsseq_6 <- vector("list", length(bsseq_6.names))
names(bsseq_6) <- bsseq_6.names
print("Empty list")
bsseq_6
count <- 1
file_paths = list()
for (i in c(73:77, 79:80)) {
  file_paths[[count]] = c(paste0("/scratch/groups/ximenac/PRJNA214817/Bismark_Aligned/Local/Meth_Extract/SRR9501", i, "_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz"))
  count <- count+1
}
#Reading in data can take up to 2-3 minutes
count <- 1
for (i in c(73:77, 79:80)) {
  bsseq_6[[count]] <- read.bismark(files = c(file_paths[[count]]),
                               colData = DataFrame(row.names = paste0(("SRR9501"), i)) ,
                                  rmZeroCov = FALSE, 
                                  strandCollapse = TRUE, 
                                  verbose = TRUE)
  count <- count+1
}

#Removing alternative loci and haplotypes if any
chr_size <- seqlengths(BSgenome.Mmusculus.UCSC.mm10)
chr_size <- chr_size[1:22]
names(chr_size) <- gsub('chr','',names(chr_size))
names(chr_size)[names(chr_size)=='M'] <- 'MT' #for Ensmbl "1" and not "chr1"

for (i in 1:length(bsseq_6)) {
  bsseq_6[[i]] <- sort(bsseq_6[[i]])
  chr_size1 <- chr_size[match(names(seqlengths(bsseq_6[[i]])),names(chr_size))[!is.na(match(names(seqlengths(bsseq_6[[i]])),names(chr_size)))]]
  bsseq_6[[i]] <- bsseq_6[[i]][seqnames(bsseq_6[[i]]) %in% names(chr_size1),]
  seqlevels(bsseq_6[[i]]) <- names(chr_size1)
  seqlengths(bsseq_6[[i]]) <- chr_size1
}

#Getting Methylation coverage to test if data that was read in is accurate
regions <- GRanges(seqnames = c("18", "19"), 
                   ranges = IRanges(start = 3.9 * 10^7 + c(0,400000), 
                                    width = 1000))
getCoverage(bsseq_6[[3]], regions = regions, what = "perBase")
bsseq_6
saveRDS(bsseq_6, ('/home/groups/ximenac/XCD_BSSeq/PRJNA294311/PRJNA294311.bsseq.rds'))
```


## 1.1 Smoothing 1. WT Mouse whole Embryo E8.5 dataset (PRJNA541237) ##

```{r}
#Warning: Only use multicore=6 when you have requested more than 6 cores from sherlock to start the interactive r session
mcp <- MulticoreParam(workers=6, progressbar = TRUE,log = TRUE)

bsseq_1 <- readRDS("/home/groups/ximenac/XCD_BSSeq/PRJNA541237/PRJNA541237.bsseq.rds")

bis.smoothed <- list()
for (i in 1: length(bsseq_1)) {
  bis.smoothed[[i]] <- BSmooth(BSseq = bsseq_1[[i]],
                               BPPARAM = mcp,
                        ns = 70,
                        h = 1000,
                        verbose=TRUE)
  bis.smoothed[[i]] <- bis.smoothed[[i]][getCoverage(bis.smoothed[[i]], type='Cov') > 1,]
saveRDS(bis.smoothed[[i]], file=paste0(i, 'PRJNA541237_smoothed.rds'))
}

#Getting Methylation coverage to test if data that was read in is accurate
regions <- GRanges(seqnames = c("18", "19"), 
                   ranges = IRanges(start = 3.9 * 10^7 + c(0,400000), 
                                    width = 1000))
getCoverage(bis.smoothed[[1]], regions = regions, what = "perBase")
bis.smoothed
```

### 2.1 Smoothing Mouse Embryo Retina Dataset E11.5, E12.5 (PRJNA521997)
```{r}
#Warning: Only use multicore=6 when you have requested more than 6 cores from sherlock to start the interactive r session
mcp <- MulticoreParam(workers=6, progressbar = TRUE,log = TRUE)

bsseq_2 <- readRDS("/home/groups/ximenac/XCD_BSSeq/PRJNA521997/PRJNA521997.bsseq.rds")

bis.smoothed <- list()
for (i in 1:length(bsseq_2)) {
  bis.smoothed[[i]] <- BSmooth(BSseq = bsseq_2[[i]],
                               BPPARAM = mcp,
                        ns = 70,
                        h = 1000,
                        verbose=TRUE)
  bis.smoothed[[i]] <- bis.smoothed[[i]][getCoverage(bis.smoothed[[i]], type='Cov') > 1,]
saveRDS(bis.smoothed[[i]], file=paste0(i, 'PRJNA521997_smoothed.rds'))
}

#Getting Methylation coverage to test if data that was read in is accurate
regions <- GRanges(seqnames = c("18", "19"), 
                   ranges = IRanges(start = 3.9 * 10^7 + c(0,400000), 
                                    width = 1000))
getCoverage(bis.smoothed[[1]], regions = regions, what = "perBase")
bis.smoothed
```


### 3.1 Smoothing Embryo & Neonate Retinae E14.5, E17.5, P3 (PRJNA343691)
```{r}
#Warning: Only use multicore=6 when you have requested more than 6 cores from sherlock to start the interactive r session
mcp <- MulticoreParam(workers=6, progressbar = TRUE,log = TRUE)

bsseq_3 <- readRDS("/home/groups/ximenac/XCD_BSSeq/PRJNA343691/PRJNA343691.bsseq.rds")

bis.smoothed <- list()
for(i in 1:length(bsseq_3)) {
  bis.smoothed[[i]] <- BSmooth(BSseq = bsseq_3[[i]],
                               BPPARAM = mcp,
                        ns = 70,
                        h = 1000,
                        verbose=TRUE)
  bis.smoothed[[i]] <- bis.smoothed[[i]][getCoverage(bis.smoothed[[i]], type='Cov') > 1,]
  saveRDS(bis.smoothed[[i]], file=paste0(i, '_PRJNA343691_smoothed.rds'))
}

#Getting Methylation coverage to test if data that was read in is accurate
regions <- GRanges(seqnames = c("18", "19"), 
                   ranges = IRanges(start = 3.9 * 10^7 + c(0,400000), 
                                    width = 1000))
getCoverage(bis.smoothed[[1]], regions = regions, what = "perBase")
bis.smoothed
```



```{r}

```



```{r}

```



```{r}

```


#### Smoothing - Rod Photoreceptor
```{r Running smoothing - Rod Photoreceptor}
#Warning: Only use multicore=6 when you have requested more than 6 cores from sherlock to start the interactive r session
mcp <- MulticoreParam(workers=6, progressbar = TRUE,log = TRUE)

bsseq_5 <- readRDS("/home/groups/ximenac/XCD_BSSeq/GSE134873/GSE134873.bsseq.rds")

#bis.smoothed = list()
bis.smoothed5.1 <- BSmooth(BSseq = bsseq_5$bsseq_5.1,
                               BPPARAM = mcp,
                        ns = 70,
                        h = 1000,
                        verbose=TRUE)
bis.smoothed5.1 <- bis.smoothed1.1[getCoverage(bis.smoothed5.1, type='Cov') > 1,]
saveRDS(bis.smoothed5.1, file=('GSE134873_62_smoothed.rds'))


bis.smoothed5.2 <- BSmooth(BSseq = bsseq_5$bsseq_5.2,
                               BPPARAM = mcp,
                        ns = 70,
                        h = 1000,
                        verbose=TRUE)
bis.smoothed5.2 <- bis.smoothed1.2[getCoverage(bis.smoothed5.2, type='Cov') > 1,]
saveRDS(bis.smoothed5.2, file=('GSE134873_63_smoothed.rds'))

bis.smoothed5.3 <- BSmooth(BSseq = bsseq_5$bsseq_5.3,
                               BPPARAM = mcp,
                        ns = 70,
                        h = 1000,
                        verbose=TRUE)
bis.smoothed5.3 <- bis.smoothed5.3[getCoverage(bis.smoothed5.3, type='Cov') > 1,]
saveRDS(bis.smoothed5.3, file=('GSE134873_64_smoothed.rds'))

```


```{r}

```



#### WT E8.5 Embryo dataset Smoothing
```{r Smoothing}
bsseq_1 <- readRDS("/home/groups/ximenac/XCD_BSSeq/PRJNA541237/PRJNA541237.bsseq.rds")
mcp <- MulticoreParam(workers=6, progressbar = TRUE,log = TRUE)

bis.smoothed1 = list()

for (i in 1:length(bsseq_1)) {
bis.smoothed1[[i]] <- BSmooth(BSseq = bsseq_1[[i]],
                               BPPARAM = mcp,
                        ns = 70,
                        h = 1000,
                        verbose=TRUE)
bis.smoothed1[[i]] <- bis.smoothed1[[i]][getCoverage(bis.smoothed1[[i]], type='Cov') > 1,]
saveRDS(bis.smoothed1[[i]], file=paste0(i,'PRJNA541237_smoothed.rds'))
}


```


#### HSC Dataset Smoothing
```{r Run the smoothing algorithm}

#Running smoothing

#Warning: Only use multicore=6 when you have requested more than 6 cores from sherlock to start the interactive r session
#bsseq3 <- readRDS("/home/groups/ximenac/XCD_BSSeq/PRJNA214817/PRJNA214817.bsseq.rds")

multicoreParam <- MulticoreParam(workers = 6, progressbar=TRUE, log=TRUE)

bis.smoothed3.1 <- BSmooth(BSseq = bsseq_3.1,
                        ns = 70,
                        h = 1000,
                        BPPARAM = multicoreParam,
                        verbose=TRUE)
bis.smoothed3.1 <- bis.smoothed3.1[getCoverage(bis.smoothed3.1, type='Cov') > 1,]
saveRDS(bis.smoothed3.1, file=('PRJNA214817_73_smoothed.rds'))

bis.smoothed3.2 <- BSmooth(BSseq = bsseq_3.2,
                        ns = 70,
                        h = 1000,
                        BPPARAM = multicoreParam,
                        verbose=TRUE)
bis.smoothed3.2 <- bis.smoothed3.2[getCoverage(bis.smoothed3.2, type='Cov') > 1,]
saveRDS(bis.smoothed3.2, file=('PRJNA214817_74_smoothed.rds'))
```

