---
title: "BSSeq_smoothing"
author: "Shruti Singh Kakan"
date: "2025-03-26"
output: html_document
---

```{r setup, include=FALSE}

#For operating this script on Sherlock, Library installations must be done in Unix Shell. 
knitr::opts_chunk$set(echo = TRUE)

```

#### Loading libraries and other dependencies
```{r Loading dependencies, message=FALSE, warning=FALSE, paged.print=FALSE}
.libPaths('/home/users/singhkak/R/x86_64-pc-linux-gnu-library/4.2')
#library(bsseq,lib.loc='~/R/3.6/library')
library(rtracklayer, lib.loc = '~/R/x86_64-pc-linux-gnu-library/4.2')
library(bsseq)
library(BSgenome.Mmusculus.UCSC.mm10)
#library(BSgenome.Hsapiens.UCSC.hg38)
library(BiocParallel)
library(parallel)

options(MulticoreParam=MulticoreParam(workers=4))
```

#### Chromosome information
```{r}
chr_size <- seqlengths(BSgenome.Mmusculus.UCSC.mm10)
#Removing alternative loci and haplotypes
chr_size <- chr_size[1:22]
#names(chr_size) <- gsub('chr','',names(chr_size))
#names(chr_size)[names(chr_size)=='M'] <- 'MT'
```

#### Rod photoreceptor datasets
```{r Reading in samples from GSE dataset}
#Mouse Rod Photo-receptors dataset
setwd("/scratch/groups/ximenac/GSE134873/Bismark_Aligned/Local/Meth_Extract")

#Reading in data can take up to 2-3 minutes
bsseq_1.1 <- read.bismark(files = c('/scratch/groups/ximenac/GSE134873/Bismark_Aligned/Local/Meth_Extract/SRR9833662_1_val_1_val_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz'), 
                                  colData = DataFrame(row.names = c("SRR9833662"),
                                  rmZeroCov = FALSE, 
                                  strandCollapse = TRUE, 
                                  verbose = TRUE))
bsseq_1.2 <- read.bismark(files = c('/scratch/groups/ximenac/GSE134873/Bismark_Aligned/Local/Meth_Extract/SRR9833663_1_val_1_val_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz'), 
                                  colData = DataFrame(row.names = c("SRR9833663"),
                                  rmZeroCov = FALSE, 
                                  strandCollapse = TRUE, 
                                  verbose = TRUE))

bsseq_1.3 <- read.bismark(files = c('/scratch/groups/ximenac/GSE134873/Bismark_Aligned/Local/Meth_Extract/SRR9833664_1_val_1_val_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz'), 
                                  colData = DataFrame(row.names = c("SRR9833664"),
                                  rmZeroCov = FALSE, 
                                  strandCollapse = TRUE, 
                                  verbose = TRUE))

chr_size <- seqlengths(BSgenome.Mmusculus.UCSC.mm10)
#Removing alternative loci and haplotypes
chr_size <- chr_size[1:22]

#Reducing BSseq object to 19 chromosomes, X, Y and MT
for (i in 1:length(bsseq1)) {
  bsseq1[[i]] <- sort(bsseq1[[i]])
  chr_size1 <- chr_size[match(names(seqlengths(bsseq1[[i]])),names(chr_size))[!is.na(match(names(seqlengths(bsseq1[[i]])),names(chr_size)))]]
  bsseq1[[i]] <- bsseq1[[i]][seqnames(bsseq1[[i]]) %in% names(chr_size1),]
  seqlevels(bsseq1[[i]]) <- names(chr_size1)
  seqlengths(bsseq1[[i]]) <- chr_size1
  seqlevels(bsseq1[[i]]) <- seqlevels(bsseq_2.1)  # align seqlevels
  seqinfo(bsseq1[[i]]) <- seqinfo(bsseq_2.1)  # guaranteed to work to change "chr1" to "1"
  print(head(seqnames(bsseq1[[i]]), n = 4))
  print(dim(bsseq1[[i]]))
}

#Getting Methylation coverage
regions <- GRanges(seqnames = c("18", "19"), 
                   ranges = IRanges(start = 3.9 * 10^7 + c(0,400000), 
                                    width = 1000))
getCoverage(bsseq1$bsseq_1.3, regions = regions, what = "perBase")

saveRDS(bsseq1, ('/home/groups/ximenac/XCD_BSSeq/GSE134873/GSE134873.bsseq.rds'))
```

#### WT Mouse whole Embryo dataset
```{r Read in WT Embryo dataset}

names(chr_size) <- gsub('chr','',names(chr_size))
names(chr_size)[names(chr_size)=='M'] <- 'MT'

bsseq_2.1 <- read.bismark(files = c('/scratch/groups/ximenac/PRJNA541237/Bismark_Aligned/Local/Meth_Extract/SRR9016926_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz'), 
                                  colData = DataFrame(row.names = c("SRR9016926")),
                                  rmZeroCov = FALSE, 
                                  strandCollapse = TRUE, 
                                  verbose = TRUE)

bsseq_2.2 <- read.bismark(files = c('/scratch/groups/ximenac/PRJNA541237/Bismark_Aligned/Local/Meth_Extract/SRR11806587_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz'), 
                                  colData = DataFrame(row.names = c("SRR1180658")),
                                  rmZeroCov = FALSE, 
                                  strandCollapse = TRUE, 
                                  verbose = TRUE)

bsseq2 <- list(bsseq2.1=bsseq_2.1, bsseq2.2=bsseq_2.2)

for (i in 1:length(bsseq2)) {
  bsseq_2.1 <- sort(bsseq_2.1)
  chr_size1 <- chr_size[match(names(seqlengths(bsseq_2.1)),names(chr_size))[!is.na(match(names(seqlengths(bsseq_2.1)),names(chr_size)))]]
  bsseq_2.1 <- bsseq_2.1[seqnames(bsseq_2.1) %in% names(chr_size1),]
  seqlevels(bsseq_2.1) <- names(chr_size1)
  seqlengths(bsseq_2.1) <- chr_size1
}

saveRDS(bsseq2, ('/home/groups/ximenac/XCD_BSSeq/PRJNA541237/PRJNA541237.bsseq.rds'))

```

### HSC Dataset
```{r}
bsseq_3.1 <- read.bismark(files = c('/scratch/groups/ximenac/PRJNA214817/Bismark_Aligned/Local/Meth_Extract/SRR950173_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz'), 
                                  colData = DataFrame(row.names = c("SRR950173")),
                                  rmZeroCov = FALSE, 
                                  strandCollapse = TRUE, 
                                  verbose = TRUE)
bsseq_3.2 <- read.bismark(files = c('/scratch/groups/ximenac/PRJNA214817/Bismark_Aligned/Local/Meth_Extract/SRR950174_1_bismark_bt2_pe.deduplicated.CpG_report.txt.gz'), 
                                  colData = DataFrame(row.names = c("SRR950174")),
                                  rmZeroCov = FALSE, 
                                  strandCollapse = TRUE, 
                                  verbose = TRUE)

#Removing alternative loci and haplotypes if any
chr_size <- seqlengths(BSgenome.Mmusculus.UCSC.mm10)
chr_size <- chr_size[1:22]
names(chr_size) <- gsub('chr','',names(chr_size))
names(chr_size)[names(chr_size)=='M'] <- 'MT'

#Getting Methylation coverage
regions <- GRanges(seqnames = c("19", "19"), 
                   ranges = IRanges(start = 3.938 * 10^7 + c(0,120000), 
                                    width = 1000))

bsseq3 <- list(bsseq_3.1=bsseq_3.1, bsseq_3.2=bsseq_3.2)
for (i in 1:length(bsseq3)) {
  bsseq3[[i]] <- sort(bsseq3[[i]])
  chr_size1 <- chr_size[match(names(seqlengths(bsseq3[[i]])),names(chr_size))[!is.na(match(names(seqlengths(bsseq3[[i]])),names(chr_size)))]]
  bsseq3[[i]] <- bsseq3[[i]][seqnames(bsseq3[[i]]) %in% names(chr_size1),]
  seqlevels(bsseq3[[i]]) <- names(chr_size1)
  seqlengths(bsseq3[[i]]) <- chr_size1
  print(getCoverage(bsseq3[[i]], regions = regions, what = "perBase"))
}

#Warning: GRanges object contains 36519 out-of-bound ranges located on sequences 4, 8, 13, 14, and 17. Note that ranges located on a sequence whose length is unknown (NA) or on a circular sequence are not considered out-of-bound (use seqlengths() and isCircular() to get the lengths and circularity flags of the underlying sequences). You can use trim() to trim these ranges. See ?`trim,GenomicRanges-method` for more information.

#saveRDS(bsseq_3, ('/home/groups/ximenac/XCD_BSSeq/PRJNA214817/PRJNA214817.bsseq.rds'))


```


#### Smoothing - Rod Photoreceptor
```{r Running smoothing - Rod Photoreceptor}
#Warning: Only use multicore=6 when you have requested more than 6 cores from sherlock to start the interactive r session
mcp <- MulticoreParam(workers=6, progressbar = TRUE,log = TRUE)

bsseq1 <- readRDS("/home/groups/ximenac/XCD_BSSeq/GSE134873/GSE134873.bsseq.rds")

#bis.smoothed = list()
bis.smoothed1.1 <- BSmooth(BSseq = bsseq1$bsseq_1.1,
                               BPPARAM = mcp,
                        ns = 70,
                        h = 1000,
                        verbose=TRUE)
bis.smoothed1.1 <- bis.smoothed1.1[getCoverage(bis.smoothed1.1, type='Cov') > 1,]
saveRDS(bis.smoothed1.1, file=('GSE134873_62_smoothed.rds'))


bis.smoothed1.2 <- BSmooth(BSseq = bsseq1$bsseq_1.2,
                               BPPARAM = mcp,
                        ns = 70,
                        h = 1000,
                        verbose=TRUE)
bis.smoothed1.2 <- bis.smoothed1.2[getCoverage(bis.smoothed1.2, type='Cov') > 1,]
saveRDS(bis.smoothed1.2, file=('GSE134873_63_smoothed.rds'))

bis.smoothed1.3 <- BSmooth(BSseq = bsseq1$bsseq_1.3,
                               BPPARAM = mcp,
                        ns = 70,
                        h = 1000,
                        verbose=TRUE)
bis.smoothed1.3 <- bis.smoothed1.3[getCoverage(bis.smoothed1.3, type='Cov') > 1,]
saveRDS(bis.smoothed1.3, file=('GSE134873_64_smoothed.rds'))

```


```{r}

```



##### WT E8.5 Embryo dataset Smoothing
```{r Smoothing}
bsseq2 <- readRDS("/home/groups/ximenac/XCD_BSSeq/PRJNA541237/PRJNA541237.bsseq.rds")
mcp <- MulticoreParam(workers=6, progressbar = TRUE,log = TRUE)

bis.smoothed2 = list()

for (i in 1:length(bsseq2)) {
bis.smoothed2[[i]] <- BSmooth(BSseq = bsseq2[[i]],
                               BPPARAM = mcp,
                        ns = 70,
                        h = 1000,
                        verbose=TRUE)
bis.smoothed2[[i]] <- bis.smoothed2[[i]][getCoverage(bis.smoothed2[[i]], type='Cov') > 1,]
saveRDS(bis.smoothed2[[i]], file=paste0(i,'PRJNA54_smoothed.rds'))
}


```


##### HSC Dataset Smoothing
```{r Run the smoothing algorithm}

#Running smoothing

#Warning: Only use multicore=6 when you have requested more than 6 cores from sherlock to start the interactive r session
#bsseq3 <- readRDS("/home/groups/ximenac/XCD_BSSeq/PRJNA214817/PRJNA214817.bsseq.rds")

multicoreParam <- MulticoreParam(workers = 6, progressbar=TRUE, log=TRUE)

bis.smoothed3.1 <- BSmooth(BSseq = bsseq_3.1,
                        ns = 70,
                        h = 1000,
                        BPPARAM = multicoreParam,
                        verbose=TRUE)
bis.smoothed3.1 <- bis.smoothed3.1[getCoverage(bis.smoothed3.1, type='Cov') > 1,]
saveRDS(bis.smoothed3.1, file=('PRJNA214817_73_smoothed.rds'))

bis.smoothed3.2 <- BSmooth(BSseq = bsseq_3.2,
                        ns = 70,
                        h = 1000,
                        BPPARAM = multicoreParam,
                        verbose=TRUE)
bis.smoothed3.2 <- bis.smoothed3.2[getCoverage(bis.smoothed3.2, type='Cov') > 1,]
saveRDS(bis.smoothed3.2, file=('PRJNA214817_74_smoothed.rds'))
```

